nx/lib/nx/defn/composite.ex

```diff
@@ -192,8 +193,12 @@ defmodule Nx.Defn.Composite do
   def to_inputs(args) do
     {args, _} =
       Enum.map_reduce(args, 0, fn arg, i ->
-        traverse(arg, i, fn arg, i ->
-          {Expr.parameter(Nx.to_tensor(arg), :root, i), i + 1}
+        traverse(arg, i, fn
+          arg, _i when is_boolean(arg) ->
+            raise ArgumentError, "booleans are not supported as defn inputs"
```

Comment:
I've added this for better error messages when booleans are passed as inputs. Since defns are now aware of booleans, I thought it was better than letting a ProtocolUndefinedError raise for Nx.Container

Reply:
This is also a point where we could decide to support booleans as defn inputs through calling Nx.tensor on them as well, though